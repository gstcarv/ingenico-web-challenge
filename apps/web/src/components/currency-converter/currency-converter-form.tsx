import { Button, DatePicker, Field } from "@ingenico-challenge/ui";
import { ArrowLeftRight, ArrowUpDown, Calendar } from "lucide-react";
import { CurrencySelect } from "./currency-select";
import { CurrencyAmountInput } from "./currency-amount-input";
import { ConvertButton } from "./convert-button";
import { useCurrencyForm } from "../../hooks/use-currency-form";
import { Controller } from "react-hook-form";

export const CurrencyConverterForm = () => {
    const { form, onSubmit, handleSwapCurrencies, handleRetry, isSubmitting, errors, conversionError } =
        useCurrencyForm();

    return (
        <form onSubmit={onSubmit} className="flex flex-col gap-6">
            <div className="flex flex-col md:flex-row items-center justify-center gap-1 md:gap-6">
                <CurrencySelect
                    label="From"
                    name="fromCurrency"
                    control={form.control}
                    error={errors.fromCurrency?.message}
                />

                <Button
                    variant="outline"
                    size="sm"
                    className="w-10 h-10 rounded-full mt-5"
                    type="button"
                    onClick={handleSwapCurrencies}
                    aria-label="Swap currencies"
                >
                    <ArrowLeftRight className="hidden lg:block" aria-hidden="true" />
                    <ArrowUpDown className="lg:hidden" aria-hidden="true" />
                </Button>

                <CurrencySelect
                    label="To"
                    name="toCurrency"
                    control={form.control}
                    error={errors.toCurrency?.message}
                />
            </div>

            <Field>
                <Field.Label htmlFor="amount">Value</Field.Label>
                <CurrencyAmountInput
                    control={form.control}
                    suffix={
                        <Controller
                            name="date"
                            control={form.control}
                            render={({ field: dateField }) => (
                                <DatePicker
                                    value={dateField.value}
                                    onChange={dateField.onChange}
                                    renderInput={({ onClick, value }) => (
                                        <Button variant="outline" size="sm" type="button" onClick={onClick}>
                                            <Calendar className="size-4 mr-2" /> {value || "Select a date"}
                                        </Button>
                                    )}
                                />
                            )}
                        />
                    }
                />

                {errors.amount && <Field.Validation>{errors.amount.message}</Field.Validation>}
            </Field>

            <div className="pt-2">
                <ConvertButton
                    isSubmitting={isSubmitting}
                    hasError={!!conversionError}
                    onRetry={handleRetry}
                    onSubmit={onSubmit}
                />
            </div>
        </form>
    );
};
