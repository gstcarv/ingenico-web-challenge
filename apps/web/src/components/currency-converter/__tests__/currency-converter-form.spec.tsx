import { render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { CurrencyConverterForm } from "../currency-converter-form";

// Mock child components
vi.mock("../currency-select", () => ({
    CurrencySelect: ({ label, name }: { label: string; name: string }) => (
        <div data-testid={`currency-select-${name}`}>
            {label} Currency Select ({name})
        </div>
    ),
}));

vi.mock("../currency-amount-input", () => ({
    CurrencyAmountInput: () => <div data-testid="currency-amount-input">Currency Amount Input</div>,
}));

vi.mock("../convert-button", () => ({
    ConvertButton: () => <div data-testid="convert-button">Convert Button</div>,
}));

// Mock the hook
const mockOnSubmit = vi.fn();
const mockHandleSwapCurrencies = vi.fn();
const mockHandleRetry = vi.fn();
const mockHandleSubmit = vi.fn((fn) => {
    // Return a function that calls the original function and our mock
    return (event: React.FormEvent) => {
        event?.preventDefault?.();
        fn();
        mockOnSubmit();
    };
});

vi.mock("../../../hooks/use-currency-form", () => ({
    useCurrencyForm: () => ({
        form: {
            control: {},
            handleSubmit: mockHandleSubmit,
        },
        onSubmit: mockOnSubmit,
        handleSwapCurrencies: mockHandleSwapCurrencies,
        handleRetry: mockHandleRetry,
        isSubmitting: false,
        errors: {},
        conversionError: null,
    }),
}));

const TestWrapper = ({ children }: { children: React.ReactNode }) => {
    const queryClient = new QueryClient({
        defaultOptions: {
            queries: { retry: false },
            mutations: { retry: false },
        },
    });

    return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

describe("CurrencyConverterForm", () => {
    it("should render the component with correct structure", () => {
        const { container } = render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        expect(container.querySelector("form")).toBeInTheDocument();
    });

    it("should render currency select components", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        expect(screen.getByTestId("currency-select-fromCurrency")).toBeInTheDocument();
        expect(screen.getByTestId("currency-select-toCurrency")).toBeInTheDocument();
    });

    it("should render swap button", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        expect(screen.getByRole("button")).toBeInTheDocument();
    });

    it("should render currency amount input", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        expect(screen.getByTestId("currency-amount-input")).toBeInTheDocument();
    });

    it("should render convert button", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        expect(screen.getByTestId("convert-button")).toBeInTheDocument();
    });

    it("should render field labels", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        expect(screen.getByText("Value")).toBeInTheDocument();
    });

    it("should have correct form structure", () => {
        const { container } = render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        const form = container.querySelector("form");
        expect(form).toHaveClass("flex", "flex-col", "gap-6");
    });

    it("should match snapshot", () => {
        const { container } = render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("should call handleSwapCurrencies when swap button is clicked", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        const swapButton = screen.getByRole("button");
        swapButton.click();

        expect(mockHandleSwapCurrencies).toHaveBeenCalledTimes(1);
    });

    it("should call handleRetry when retry is triggered", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        // Verify that the retry function is available from the hook
        expect(mockHandleRetry).toBeDefined();
    });

    it("should pass form control to child components", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        // Verify that the form control is being used by checking if child components are rendered
        expect(screen.getByTestId("currency-select-fromCurrency")).toBeInTheDocument();
        expect(screen.getByTestId("currency-select-toCurrency")).toBeInTheDocument();
        expect(screen.getByTestId("currency-amount-input")).toBeInTheDocument();
    });

    it("should pass loading state to ConvertButton", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        // Verify that the ConvertButton is rendered and receives the isSubmitting prop
        expect(screen.getByTestId("convert-button")).toBeInTheDocument();
    });

    it("should have access to form submission function from hook", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        // Verify that the onSubmit function is available from the hook
        expect(mockOnSubmit).toBeDefined();
    });

    it("should have access to swap currencies function from hook", () => {
        render(
            <TestWrapper>
                <CurrencyConverterForm />
            </TestWrapper>,
        );

        // Verify that the handleSwapCurrencies function is available from the hook
        expect(mockHandleSwapCurrencies).toBeDefined();
    });
});
