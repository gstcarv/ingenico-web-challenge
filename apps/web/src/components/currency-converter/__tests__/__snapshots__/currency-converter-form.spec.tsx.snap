// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CurrencyConverterForm > should match snapshot 1`] = `
<form
  class="flex flex-col gap-6"
>
  <div
    class="flex flex-col md:flex-row items-center justify-center gap-1 md:gap-6"
  >
    <div
      data-testid="currency-select-fromCurrency"
    >
      From
       Currency Select (
      fromCurrency
      )
    </div>
    <button
      aria-label="Swap currencies"
      class="inline-flex items-center justify-center font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer border border-neutral-200 bg-white text-neutral-900 hover:bg-neutral-50 focus-visible:ring-neutral-500 px-3 text-sm w-10 h-10 rounded-full mt-5"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-arrow-left-right hidden lg:block"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 3 4 7l4 4"
        />
        <path
          d="M4 7h16"
        />
        <path
          d="m16 21 4-4-4-4"
        />
        <path
          d="M20 17H4"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="lucide lucide-arrow-up-down lg:hidden"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21 16-4 4-4-4"
        />
        <path
          d="M17 20V4"
        />
        <path
          d="m3 8 4-4 4 4"
        />
        <path
          d="M7 4v16"
        />
      </svg>
    </button>
    <div
      data-testid="currency-select-toCurrency"
    >
      To
       Currency Select (
      toCurrency
      )
    </div>
  </div>
  <div
    class="field flex flex-col gap-2"
  >
    <label
      class="text-xs text-neutral-500 uppercase tracking-wide font-medium"
      for="amount"
    >
      Value
    </label>
    <div
      data-testid="currency-amount-input"
    >
      Currency Amount Input
    </div>
  </div>
  <div
    class="pt-2"
  >
    <div
      data-testid="convert-button"
    >
      Convert Button
    </div>
  </div>
</form>
`;
